### PR TITLE
Enable gRPC command to support oneof using optional fields

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/20485

## Approach
- Use a map of field list to push data to handlebars
- Update the mustache templates to generate the ballerina files accordingly

## Samples

For a given protobuf file like below,

```proto
syntax = "proto3";

message Student {
	string name = 1;
	oneof identity {
		ID id = 2;
		Passport passport = 3;
	}
}

message ID {
	string idType = 7;
	string idNumber = 8;
};

message Passport {
	string ppNumber = 9;
	string country = 10;
};
```

it generates the following ballerina stub file.

```ballerina

public type Student record {|
    string name = "";
    ID id?;
    Passport passport?;
    
|};

function isValidStudent(Student r) returns boolean {
    int identityCount = 0;
    if !(r?.id is ()) {
        identityCount += 1;
    }
    if !(r?.passport is ()) {
        identityCount += 1;
    }
    
    if (identityCount > 1 ) {
        return false;
    }
    return true;
}

function setStudent_Id(Student r, ID id) {
    r.id = id;
    r.remove("passport");
    
}
function setStudent_Passport(Student r, Passport passport) {
    r.passport = passport;
    r.remove("id");
    
}


public type ID record {|
    string idType = "";
    string idNumber = "";
    
|};

public type Passport record {|
    string ppNumber = "";
    string country = "";
    
|};


const string ROOT_DESCRIPTOR = "0A126F6E656F662D73616D706C652E70726F746F22690A0753747564656E7412120A046E616D6518012001280952046E616D6512150A02696418022001280B32032E494448005202696412270A0870617373706F727418032001280B32092E50617373706F72744800520870617373706F7274420A0A086964656E7469747922380A02494412160A066964547970651807200128095206696454797065121A0A0869644E756D626572180820012809520869644E756D62657222400A0850617373706F7274121A0A0870704E756D626572180920012809520870704E756D62657212180A07636F756E747279180A200128095207636F756E747279620670726F746F33";
function getDescriptorMap() returns map<string> {
    return {
        "oneof-sample.proto":"0A126F6E656F662D73616D706C652E70726F746F22690A0753747564656E7412120A046E616D6518012001280952046E616D6512150A02696418022001280B32032E494448005202696412270A0870617373706F727418032001280B32092E50617373706F72744800520870617373706F7274420A0A086964656E7469747922380A02494412160A066964547970651807200128095206696454797065121A0A0869644E756D626572180820012809520869644E756D62657222400A0850617373706F7274121A0A0870704E756D626572180920012809520870704E756D62657212180A07636F756E747279180A200128095207636F756E747279620670726F746F33"
        
    };
}
```
## Remarks
This feature (https://github.com/ballerina-platform/ballerina-lang/issues/18873) should be supported before actually use the generated stub file.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: stdlib/grpc/src/main/resources/templates/skeleton/message.mustache [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+{{#each oneofFieldMap}}{{#each this}}
+function set{{messageName}}_{{{camelcase fieldName}}}({{messageName}} r, {{{fieldType}}} {{{fieldName}}}) {
+    r.{{{fieldName}}} = {{{fieldName}}};
+    {{#each ../this}}{{#not_equal fieldName ../fieldName}}r.remove("{{fieldName}}");
+    {{/not_equal}}{{/each}}
+}{{/each}}
+{{/each}}
+{{/if}}{{#enumList}}
 {{> enum}}{{/enumList}}{{#if this.nestedMessageList}}{{#each this.nestedMessageList}}
 {{> message}}{{/each}}{{/if}}
```
Comment (daneshk) [inline stdlib/grpc/src/main/resources/templates/skeleton/message.mustache]:
```suggestion
{{> message}}{{/each}}{{/if}}

```

File: stdlib/grpc/src/main/resources/templates/skeleton/message.mustache [hunk 2]
```diff
@@ -1,10 +1,29 @@
...
+    }
+    return true;
+}
+{{#each oneofFieldMap}}{{#each this}}
+function set{{messageName}}_{{{camelcase fieldName}}}({{messageName}} r, {{{fieldType}}} {{{fieldName}}}) {
+    r.{{{fieldName}}} = {{{fieldName}}};
+    {{#each ../this}}{{#not_equal fieldName ../fieldName}}r.remove("{{fieldName}}");
+    {{/not_equal}}{{/each}}
+}{{/each}}
+{{/each}}
```
Comment (daneshk) [inline stdlib/grpc/src/main/resources/templates/skeleton/message.mustache]: Shall we move this logic to `oneoffield.mustache` file? refer that file here

**Comment by daneshk** [comment]
@BuddhiWathsala we also need to change the protobuf encoding and decoding logic in 
https://github.com/ballerina-platform/ballerina-lang/blob/master/stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/Message.java

File: stdlib/grpc/src/main/resources/templates/skeleton/message.mustache (hunk 1 above)
Comment (BuddhiWathsala) [reply stdlib/grpc/src/main/resources/templates/skeleton/message.mustache]: fixed in https://github.com/ballerina-platform/ballerina-lang/pull/21199/commits/188fb2a1218a87545571091a498eb5512d05f926

File: stdlib/grpc/src/main/resources/templates/skeleton/message.mustache (hunk 2 above)
Comment (BuddhiWathsala) [reply stdlib/grpc/src/main/resources/templates/skeleton/message.mustache]: fixed in https://github.com/ballerina-platform/ballerina-lang/pull/21199/commits/188fb2a1218a87545571091a498eb5512d05f926

**Comment by BuddhiWathsala** [comment]
Here tests are failing due to the fact that the `remove()` function does not support in the ballerina for records, as mentioned in this issue(#18873).

**Comment by daneshk** [comment]
> @daneshk, simplify the encoding decoding logic here ([b68a133](https://github.com/ballerina-platform/ballerina-lang/commit/b68a1339f8471ff1c54073504402422121399769)). The tests are failing due to the fact that the `remove()` function does not support in the ballerina records, as mentioned in this issue(#18873).

Let's merge this once we have the compiler fix.

**Pull request merged**